server/server.go

```diff
@@ -75,34 +73,37 @@ func New(config *Config) (server *Server, err error) {
 		}
 	}
 
-	return s, nil
-}
+	r := mux.NewRouter()
+	r.HandleFunc("/", s.ping).Methods(http.MethodGet)
+	r.HandleFunc("/pr_event", s.githubEvent).Methods(http.MethodPost)
 
-// Start starts a server
-func (s *Server) Start() {
-	mlog.Info("Starting Mattermod Server")
+	s.server = &http.Server{
+		Addr:         s.Config.ListenAddress,
+		Handler:      r,
+		ReadTimeout:  60 * time.Second,
+		WriteTimeout: 60 * time.Second,
+		IdleTimeout:  30 * time.Second,
+	}
 
-	rand.Seed(time.Now().Unix())
+	return s
+}
 
-	s.initializeRouter()
+// Start starts a server
+func (s *Server) Start() <-chan error {
 	s.RefreshMembers()
 
-	var handler http.Handler = s.Router
+	errc := make(chan error)
 	go func() {
 		mlog.Info("Listening on", mlog.String("address", s.Config.ListenAddress))
-		err := manners.ListenAndServe(s.Config.ListenAddress, handler)
-		if err != nil {
-			s.logToMattermost(err.Error())
-			mlog.Critical("server_error", mlog.Err(err))
-			panic(err.Error())
-		}
+		errc <- s.server.ListenAndServe()
```

Comment:
Any reason for such a complicated implementation? `ListenAndServe` will _always_ return an error when stopped via `Shutdown`.

This means there will be deadlock now because `errc` cannot take the error since it's inside the signal handler case.

I thought we would just use the standard approach here: https://golang.org/pkg/net/http/#Server.Shutdown and just do an mlog.Error if it's not a `ErrServerClosed`.

Reply:
Hmm, actually we are implicitly ignoring `ErrServerClosed` because the interrupt signal will win in the select block in the `main` function. Otherwise we will receive errors from `ListenAndServe`. But the standard way is always more preferable :)